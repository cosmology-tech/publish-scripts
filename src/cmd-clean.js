#!/usr/bin/env node
import { clean, defaultClean } from '.';

clean(defaultClean);
