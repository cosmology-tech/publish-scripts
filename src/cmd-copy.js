#!/usr/bin/env node
import { copy, defaultCopy } from '.';

copy(defaultCopy);
