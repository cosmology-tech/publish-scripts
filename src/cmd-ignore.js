#!/usr/bin/env node
import { ignore, defaultIgnore } from '.';

ignore(defaultIgnore);
