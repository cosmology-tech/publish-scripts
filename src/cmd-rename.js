#!/usr/bin/env node
import { rename, defaultRename } from '.';

rename(defaultRename);
