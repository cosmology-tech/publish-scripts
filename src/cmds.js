import _rename from './commands/rename';
const Commands = {};
Commands['rename'] = _rename;

export { Commands };
