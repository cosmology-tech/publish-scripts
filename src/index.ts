import { globSync as glob } from 'glob';
import { mkdirpSync as mkdirp } from 'mkdirp';
import { extname, join, dirname, format, parse } from 'path';
import { copyFileSync, readdirSync, rmdirSync, statSync } from 'fs';
import { rimrafSync } from 'rimraf';

export const rename = ({
    findExt,
    replaceExt,
    outDir = join(process.cwd(), '__output__'),
    srcDir = process.cwd()
}) => {
    mkdirp(outDir);
    const files = glob('**/*.' + findExt, { ignore: 'node_modules/**', cwd: srcDir });
    files.forEach(file => {
        const src = file;
        const dst = format({ ...parse(src), base: '', ext: '.' + replaceExt });
        const outDst = join(outDir, dst);
        mkdirp(dirname(outDst));
        copyFileSync(join(srcDir, src), join(outDir, dst));
    })
};

interface CopyFn {
    findExt: string | string[];
    outDir: string;
    srcDir: string;
    stripPath: string;
}
export const copy = ({
    findExt = [],
    outDir = join(process.cwd(), '__output__'),
    srcDir = process.cwd(),
    stripPath = ''
}: CopyFn) => {
    const regexp = new RegExp('^' + stripPath);

    mkdirp(outDir);
    if (typeof findExt === 'string') findExt = findExt.split(',').map(a => a.trim()).filter(Boolean)
    const files = findExt.reduce((m, v) => {
        return [...m, ...glob('**/*.' + v, { ignore: 'node_modules/**', cwd: srcDir })];
    }, []);

    files.forEach(file => {
        let src = file;
        if (regexp.test(file)) {
            src = file.replace(regexp, '');
        }
        const outDst = join(outDir, src);
        mkdirp(dirname(outDst));
        copyFileSync(join(srcDir, file), join(outDir, src));
    })
};
interface CleanFn {
    findExt: string | string[];
    outDir: string;
    srcDir: string;
    stripPath: string;
    removeEmpty: boolean;
}


const cleanEmptyFoldersRecursively = (folder) => {
    const isDir = statSync(folder).isDirectory();
    if (!isDir) return;
    let files = readdirSync(folder);
    if (files.length > 0) {
        files.forEach((file) => {
            var fullPath = join(folder, file);
            cleanEmptyFoldersRecursively(fullPath);
        });

        // re-evaluate files; after deleting subfolder
        // we may have parent folder empty now
        files = readdirSync(folder);
    }

    if (files.length == 0) {
        rmdirSync(folder);
        return;
    }
}

// deletes the files that would have otherwise been copied
export const clean = ({
    findExt = [],
    outDir = join(process.cwd(), '__output__'),
    srcDir = process.cwd(),
    stripPath = '',
    removeEmpty = true
}: CleanFn) => {
    const regexp = new RegExp('^' + stripPath);

    mkdirp(outDir);
    if (typeof findExt === 'string') findExt = findExt.split(',').map(a => a.trim()).filter(Boolean)
    const files = findExt.reduce((m, v) => {
        return [...m, ...glob('**/*.' + v, { ignore: 'node_modules/**', cwd: srcDir })];
    }, []);

    files.forEach(file => {
        let src = file;
        if (regexp.test(file)) {
            src = file.replace(regexp, '');
        }
        const outDst = join(outDir, src);
        mkdirp(dirname(outDst));
        rimrafSync(join(outDir, src));
    });

    // clean empty dirs
    if (removeEmpty) {
        cleanEmptyFoldersRecursively(outDir);
    }
};

interface IgnoreFn {
    findExt: string | string[];
    outDir: string;
    srcDir: string;
    stripPath: string;
}

// log the files that would have otherwise been copied
export const ignore = ({
    findExt = [],
    outDir = join(process.cwd(), '__output__'),
    srcDir = process.cwd(),
    stripPath = ''
}: IgnoreFn) => {
    const regexp = new RegExp('^' + stripPath);

    if (typeof findExt === 'string') findExt = findExt.split(',').map(a => a.trim()).filter(Boolean)
    const files = findExt.reduce((m, v) => {
        return [...m, ...glob('**/*.' + v, { ignore: 'node_modules/**', cwd: srcDir })];
    }, []);

    // only get the top level files (this can prob be more efficient)
    const del = files.reduce((m, file) => {
        let src = file;
        if (regexp.test(file)) {
            src = file.replace(regexp, '');
        }
        const srcs = src.split('/').map(a => a.trim()).filter(Boolean);
        const first = srcs?.[0] ?? '';
        if (first !== '') {
            m[first] = true;
        }
        return m;
    }, {});

    return Object.keys(del).reduce((m, v) => {
        return `${m}/${v}\n`
    }, '');
};
