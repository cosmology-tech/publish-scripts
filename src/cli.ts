import { prompt } from './utils';
import { Commands as commands } from './cmds';

const question = [
    {
        _: true,
        type: 'fuzzy',
        name: 'cmd',
        message: 'what do you want to do?',
        choices: Object.keys(commands)
    }
];

export const cli = async (argv) => {
    if (argv.clean) argv.cmd = 'clean';
    if (argv.rename) argv.cmd = 'rename';
    if (argv.copy) argv.cmd = 'copy';
    if (argv.ignore) argv.cmd = 'ignore';

    var { cmd } = await prompt(question, argv);
    if (typeof commands[cmd] === 'function') {
        await commands[cmd](argv);
    } else {
        console.log('command not found.');
    }
};
