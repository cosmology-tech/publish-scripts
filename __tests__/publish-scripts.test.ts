import { rename, clean, copy, ignore } from '../src/index';
import { join, resolve } from 'path';
import { globSync as glob } from 'glob';
import { rimrafSync } from 'rimraf';

const FIXTURES = resolve(join(__dirname, '/../__fixtures__/'))
const OUTDIR = resolve(join(__dirname, '/../__output__/'))

it('rename', async () => {
    const outDir = join(OUTDIR, 'rename-test');
    const curDir = join(FIXTURES, 'rename-test');
    rimrafSync(outDir);
    rename({
        findExt: 'js',
        replaceExt: 'texta',
        srcDir: join(curDir, 'builda'),
        outDir
    });
    rename({
        findExt: 'js',
        replaceExt: 'textb',
        srcDir: join(curDir, 'buildb'),
        outDir
    });
    const a = glob(outDir + '/**/*').map(f => f.replace(outDir, '').replace(/^\//, ''))
    expect(a).toMatchSnapshot();
});

it('copy', async () => {
    const outDir = join(OUTDIR, 'copy-test');
    const curDir = join(FIXTURES, 'copy-test');
    rimrafSync(outDir);
    copy({
        findExt: 'texta,textb,tx,ty',
        srcDir: curDir,
        outDir,
        stripPath: ''
    });
    const a = glob(outDir + '/**/*').map(f => f.replace(outDir, '').replace(/^\//, ''))
    expect(a).toMatchSnapshot();
});

it('copy stripPath', async () => {
    const outDir = join(OUTDIR, 'copy-test');
    const curDir = join(FIXTURES, 'copy-test');
    rimrafSync(outDir);
    copy({
        findExt: 'texta,textb,tx,ty',
        srcDir: curDir,
        outDir,
        stripPath: 'some/path'
    });
    const a = glob(outDir + '/**/*').map(f => f.replace(outDir, '').replace(/^\//, ''))
    expect(a).toMatchSnapshot();
});

it('copy stripPath bad', async () => {
    const outDir = join(OUTDIR, 'copy-test');
    const curDir = join(FIXTURES, 'copy-test');
    rimrafSync(outDir);
    copy({
        findExt: 'texta,textb,tx,ty',
        srcDir: curDir,
        outDir,
        stripPath: 'another/some/path'
    });
    const a = glob(outDir + '/**/*').map(f => f.replace(outDir, '').replace(/^\//, ''))
    expect(a).toMatchSnapshot();
});

it('clean', async () => {
    const outDir = join(OUTDIR, 'clean-test');
    const curDir = join(FIXTURES, 'clean-test');
    rimrafSync(outDir);
    copy({
        findExt: 'texta,textb,tx,ty',
        srcDir: curDir,
        outDir,
        stripPath: ''
    });
    const a = glob(outDir + '/**/*').map(f => f.replace(outDir, '').replace(/^\//, ''))
    expect(a).toMatchSnapshot();

    clean({
        findExt: 'texta,textb,tx,ty',
        srcDir: curDir,
        outDir,
        stripPath: '',
        removeEmpty: true
    });
    const b = glob(outDir + '/**/*').map(f => f.replace(outDir, '').replace(/^\//, ''))
    expect(b).toMatchSnapshot();

});

it('ignore', async () => {
    const outDir = join(OUTDIR, 'ignore-test');
    const curDir = join(FIXTURES, 'ignore-test');
    const a = ignore({
        findExt: 'texta,textb,tx,ty',
        srcDir: curDir,
        outDir,
        stripPath: 'another/some/path'
    });
    expect(a).toMatchSnapshot();
});