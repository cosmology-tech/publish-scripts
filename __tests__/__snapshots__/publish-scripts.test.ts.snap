// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clean 1`] = `
[
  "some",
  "g.ty",
  "g.tx",
  "g.textb",
  "some/path",
  "some/path/g.ty",
  "some/path/g.tx",
  "some/path/g.textb",
  "some/path/g.texta",
  "some/path/b",
  "some/path/a",
  "some/path/b/c",
  "some/path/b/c/d",
  "some/path/b/c/d/e.textb",
  "some/path/b/c/d/e.texta",
  "some/path/a/d",
  "some/path/a/d/e.textb",
  "some/path/a/d/e.texta",
]
`;

exports[`clean 2`] = `[]`;

exports[`copy 1`] = `
[
  "some",
  "some/path",
  "some/path/g.ty",
  "some/path/g.tx",
  "some/path/g.textb",
  "some/path/g.texta",
  "some/path/b",
  "some/path/a",
  "some/path/b/c",
  "some/path/b/c/d",
  "some/path/b/c/d/e.textb",
  "some/path/b/c/d/e.texta",
  "some/path/a/d",
  "some/path/a/d/e.textb",
  "some/path/a/d/e.texta",
]
`;

exports[`copy stripPath 1`] = `
[
  "g.ty",
  "g.tx",
  "g.textb",
  "g.texta",
  "b",
  "a",
  "b/c",
  "b/c/d",
  "b/c/d/e.textb",
  "b/c/d/e.texta",
  "a/d",
  "a/d/e.textb",
  "a/d/e.texta",
]
`;

exports[`copy stripPath bad 1`] = `
[
  "some",
  "some/path",
  "some/path/g.ty",
  "some/path/g.tx",
  "some/path/g.textb",
  "some/path/g.texta",
  "some/path/b",
  "some/path/a",
  "some/path/b/c",
  "some/path/b/c/d",
  "some/path/b/c/d/e.textb",
  "some/path/b/c/d/e.texta",
  "some/path/a/d",
  "some/path/a/d/e.textb",
  "some/path/a/d/e.texta",
]
`;

exports[`ignore 1`] = `
"/some
/g.textb
/g.tx
/g.ty
"
`;

exports[`rename 1`] = `
[
  "a",
  "a/g.textb",
  "a/g.texta",
  "a/b",
  "a/b/c",
  "a/b/c/d",
  "a/b/c/d/e.textb",
  "a/b/c/d/e.texta",
]
`;
